### PR TITLE
chore: add virtual list support

### PR DESCRIPTION
**Pull request by qiuqfang**

### CODE REVIEW

**Review by himself65 — CHANGES_REQUESTED** [review]
Thanks for your contribution. But I'm thinking to refactor this component using tanstack/table or similar library

File: packages/component/src/components/page-list/use-virtual-table-height.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { useEffect, useState } from 'react';
+
+export function useVirtualTableHeight() {
+  const [virtualTableHeight, setVirtualTableHeight] = useState(0);
+  useEffect(() => {
+    const mainEl = document.querySelector('.main-container') as HTMLElement;
+    const headEl = document.querySelector('.header-container') as HTMLElement;
```
Comment (himself65) [inline packages/component/src/components/page-list/use-virtual-table-height.ts]: prohibit using the constant value

**Review by himself65 — CHANGES_REQUESTED** [review]

File: packages/component/src/components/page-list/use-virtual-table-height.ts (hunk 1 above)
Comment (qiuqfang) [reply packages/component/src/components/page-list/use-virtual-table-height.ts:7]: Okay, but for what reason?

Comment (himself65) [reply packages/component/src/components/page-list/use-virtual-table-height.ts]: not a good engineering approach. I think we pass two `ref<HTMLElemet>` in parameters is better.

And a hook function should be pure, which also means it only depends on the inputs(parameters)

Comment (qiuqfang) [reply packages/component/src/components/page-list/use-virtual-table-height.ts]: > not a good engineering approach. I think we pass two `ref<HTMLElemet>` in parameters is better.
> 
> And a hook function should be pure, which also means it only depends on the inputs(parameters)

I've modified it.

File: apps/core/src/components/blocksuite/workspace-header/header.tsx [hunk 2]
```diff
@@ -172,7 +172,7 @@ export const Header = forwardRef<HTMLDivElement, HeaderProps>((props, ref) => {
 
   return (
     <div
-      className={styles.headerContainer}
+      className={(styles.headerContainer, 'header-container')}
```
Comment (himself65) [inline apps/core/src/components/blocksuite/workspace-header/header.tsx]: use clsx

**Review by himself65 — DISMISSED** [review]

**Review by himself65 — APPROVED** [review]
LGTM

File: packages/component/src/components/page-list/mobile.tsx [hunk 3]
```diff
@@ -68,6 +70,11 @@ export const AllPageListMobileView = ({
   createNewEdgeless: () => void;
   importFile: () => void;
 }) => {
+  const virtualTableHeight = useVirtualTableHeight(
+    '.main-container',
```
Comment (himself65) [inline packages/component/src/components/page-list/mobile.tsx]: Shouldn't use css selector here. We might have multiple page list in a single web

**Review by himself65 — CHANGES_REQUESTED** [review]

File: packages/component/src/components/page-list/all-page.tsx [hunk 4]
```diff
@@ -167,12 +169,15 @@ export const PageList = ({
   getPageInfo,
   propertiesMeta,
 }: PageListProps) => {
+  const virtualTableHeight = useVirtualTableHeight(
+    '.main-container',
+    '.header-container'
```
Comment (himself65) [inline packages/component/src/components/page-list/all-page.tsx]: We dont have `header-container` classname in the code

**Review by himself65 — CHANGES_REQUESTED** [review]

File: packages/component/src/components/page-list/virtual-table.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        innerElementType={InnerElement}
+        onItemsRendered={({
+          overscanStartIndex,
+          overscanStopIndex,
+          visibleStartIndex,
+          visibleStopIndex,
+        }) => {
+          if (!listRef.current) return;
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
```
Comment (himself65) [inline packages/component/src/components/page-list/virtual-table.tsx]: Remove this

File: packages/component/src/components/page-list/virtual-table.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+const VirtualTableContext = React.createContext<{
+  header: React.ReactNode;
+  footer: React.ReactNode;
+  top: number;
+}>({
+  header: <></>,
+  footer: <></>,
+  top: 0,
+});
```
Comment (himself65) [inline packages/component/src/components/page-list/virtual-table.tsx]: use atom instead

File: packages/component/src/components/page-list/use-virtual-table-height.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import { useEffect, useState } from 'react';
+
+export function useVirtualTableHeight(
+  mainSelectors: string,
+  headSelectors: string
```
Comment (himself65) [inline packages/component/src/components/page-list/use-virtual-table-height.ts]: use `ObjectRef<HTMLElement>`

**Review by himself65 — CHANGES_REQUESTED** [review]

File: packages/component/src/components/page-list/use-virtual-table-height.ts (hunk 7 above)
Comment (himself65) [reply packages/component/src/components/page-list/use-virtual-table-height.ts]: I will handle this

**Comment by pengx17** [comment]
Hi @qiuqfang , thanks for your contribution! We are replacing the pages using `div` instead of `table` in next version and your current PR will be deprecated.
Will give credits to you in the new implementation. Thanks again!

**Pull request closed without merging**